### PR TITLE
[DCOS-46083]: Allow file and/or environment configuration

### PR DESCRIPTION
**Pull request by brandonc**
## Overview

Uses viper (https://github.com/spf13/viper) for more flexible configuration

## Testing

- You can now set config using file (--config path.json) See fixtures/config.json for example
- You can now set certain variables using system environment variables:
DCOS_UI_UPDATE_LISTEN_ADDR
DCOS_UI_UPDATE_DEFAULT_UI_PATH
DCOS_UI_UPDATE_VERSIONS_ROOT
DCOS_UI_UPDATE_DIST_LINK
DCOS_UI_UPDATE_STAGE_LINK
DCOS_UI_UPDATE_ZK_AUTH_INFO
DCOS_UI_UPDATE_ZK_ZKNODE_OWNER

## Trade-offs

Introduces a new external dependency

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -88,88 +59,166 @@ const (
...
 
+func defineFlags(viper *viper.Viper) (*pflag.FlagSet, error) {
+	fs := &pflag.FlagSet{}
+	fs.String(optConfig, defaultConfig, "The path to the optional config file")
+	fs.String(optListenNet, defaultListenNet, "The transport type on which to listen for connections. May be one of 'tcp', 'unix'.")
+	fs.String(optListenAddress, defaultListenAddr, "The network address at which to listen for connections.")
+	fs.String(optUniverseURL, defaultUniverseURL, "The URL where universe can be reached.")
+	fs.String(optDefaultDocRoot, defaultDefaultDocRoot, "The filesystem path with the default ui distribution (pre-bundled ui).")
+	fs.String(optUIDistSymlink, defaultUIDistSymlink, "The filesystem symlink path where the ui distributed files are served from.")
+	fs.String(optUIDistStageSymlink, defaultUIDistStageSymlink, "The temporary filesystem symlink path that links to where the ui distribution files are located.")
```
Comment (TattdCodeMonkey) [inline config/config.go]: linter complaining this line is > 160 characters

File: main.go [hunk 2]
```diff
@@ -13,7 +13,11 @@ import (
 // TODO: think about client timeouts https://blog.cloudflare.com/the-complete-guide-to-golang-net-http-timeouts/
 func main() {
 	cliArgs := os.Args[1:]
-	config := config.Parse(cliArgs)
+	config, err := config.Parse(cliArgs)
+
+	if err != nil {
+		logrus.WithFields(logrus.Fields{"err": err.Error()}).Fatalf("Could not load config")
```
Comment (TattdCodeMonkey) [inline main.go]: If the only field is an error you can use [logrus.WithError()](https://godoc.org/github.com/sirupsen/logrus#WithError)

File: main_test.go [hunk 3]
```diff
@@ -61,24 +61,26 @@ func listen() (net.Listener, error) {
 }
 
 func setupTestUIService() *uiService.UIService {
-	cfg := config.NewDefaultConfig()
-	cfg.DefaultDocRoot = "./testdata/main-sandbox/dcos-ui"
-	cfg.VersionsRoot = "./testdata/main-sandbox/ui-versions"
-	cfg.UIDistSymlink = "./testdata/main-sandbox/dcos-ui-dist"
-	cfg.MasterCountFile = "./fixtures/single-master"
+	cfg, _ := config.Parse([]string{
```
Comment (TattdCodeMonkey) [inline main_test.go]: 👍

File: uiService/root.go [hunk 4]
```diff
@@ -69,11 +69,11 @@ func (service *UIService) Run(l net.Listener) error {
...
 
 func checkUIDistSymlink(cfg *config.Config) {
-	uiDistTarget, err := os.Readlink(cfg.UIDistSymlink)
+	uiDistTarget, err := os.Readlink(cfg.UIDistSymlink())
 	if err != nil {
-		if cfg.InitUIDistSymlink {
+		if cfg.InitUIDistSymlink() {
 			logrus.Info("Attempting to initialize UI dist symlink")
-			createErr := os.Symlink(cfg.DefaultDocRoot, cfg.UIDistSymlink)
+			createErr := os.Symlink(cfg.DefaultDocRoot(), cfg.UIDistSymlink())
```
Comment (TattdCodeMonkey) [inline uiService/root.go]: the logs fields on lines 82, 83, 93 also need to be updated.

File: uiService/root.go [hunk 5]
```diff
@@ -114,9 +114,9 @@ func checkCurrentVersion(updateManager *updateManager.Client) {
 
 func checkVersionsRoot(cfg *config.Config) {
 	logger := logrus.WithFields(logrus.Fields{"VersionsRoot": cfg.VersionsRoot})
```
Comment (TattdCodeMonkey) [inline uiService/root.go]: `cfg.VersionRoot` -> `cfg.VersionRoot()`

**Comment by brandonc** [comment]
@TattdCodeMonkey thanks for the review. I checked all the fields to make sure they were being called as functions. I relied on the type system inappropriately!

**Comment by brandonc** [comment]
@TattdCodeMonkey hold off on merge. I found an issue with "--help" and want to update the README to document the environment config

**Comment by brandonc** [comment]
@TattdCodeMonkey OK now you can use "--help" for usage options without error and I added all the configuration docs to the README

**Pull request merged**